### ISSUE TITLE
Gollum reports encoding error on '--'

### ISSUE THREAD

**Issue by cornim** [gollum/gollum#1699]
**Describe the bug**
If a markdown file contains a double dash '--' it can not be loaded and reports an encoding error.
File is UTF-8. Removing the double dash allows rendering of the page. Writing the double dash as `\-\-` also works.

**To Reproduce**
```bash
mkdir test
cd test
git init
cat > test.md << EOF
# Test

ab --ab
EOF
git add test.md
git commit -m "test"
gollum
```
go to `localhost:4567/test.md

**Expected behavior**
Should show rendered markdown page.

**Environment Info**
```
$ gollum --versions
Gollum 5.2.1
Running on: java with Ruby version 2.5.7
Using:
rjgit 5.7.0.2
gollum-rjgit_adapter 0.6
gollum-lib 5.1
Markdown rendering gem: kramdown
Other renderering gems:
none
```

**Comment by bartkamphorst** [comment]
I'm not able to reproduce this on:
```
Gollum 5.2.1
Running on: java with Ruby version 2.6.5
Using:
rjgit 5.7.0.2
gollum-rjgit_adapter 0.6
gollum-lib 5.1
Markdown rendering gem: kramdown
Other renderering gems:
none
```

![Screenshot 2021-03-17 at 21 40 23](https://user-images.githubusercontent.com/571173/111535664-a7b67300-8769-11eb-9094-1bf31445966d.png)

**Comment by cornim** [comment]
Well that is strange. I saw the Ruby version is different in your run and I should mention that this is all executed in MINGW on Windows. But I did check that the file encoding is UTF-8.

This is the exception and stack trace being produced:
```
2021-03-18 08:33:33 - Encoding::CompatibilityError - incompatible encodings: Windows-1252 and UTF-8:
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/mustache-0.99.8/lib/mustache/template.rb:75:in `render'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/mustache-0.99.8/lib/mustache.rb:121:in `render'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/mustache-0.99.8/lib/mustache/sinatra.rb:113:in `mustache'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/gollum-5.2.1/lib/gollum/app.rb:612:in `show_page_or_file'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/gollum-5.2.1/lib/gollum/app.rb:591:in `block in GET /*'
        org/jruby/RubyMethod.java:115:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1675:in `block in compile!'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1013:in `block in route!'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1032:in `route_eval'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1013:in `block in route!'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1061:in `block in process_route'
        org/jruby/RubyKernel.java:1189:in `catch'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1059:in `process_route'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1011:in `block in route!'
        org/jruby/RubyArray.java:1809:in `each'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1008:in `route!'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1129:in `block in dispatch!'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1101:in `block in invoke'
        org/jruby/RubyKernel.java:1189:in `catch'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1101:in `invoke'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1124:in `dispatch!'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:939:in `block in call!'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1101:in `block in invoke'
        org/jruby/RubyKernel.java:1189:in `catch'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1101:in `invoke'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:939:in `call!'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:929:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-protection-2.1.0/lib/rack/protection/xss_header.rb:18:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-protection-2.1.0/lib/rack/protection/path_traversal.rb:16:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-protection-2.1.0/lib/rack/protection/json_csrf.rb:26:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-protection-2.1.0/lib/rack/protection/base.rb:50:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-protection-2.1.0/lib/rack/protection/base.rb:50:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-protection-2.1.0/lib/rack/protection/frame_options.rb:31:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-2.2.3/lib/rack/null_logger.rb:11:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-2.2.3/lib/rack/head.rb:12:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/show_exceptions.rb:22:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:216:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1991:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1542:in `block in call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1769:in `synchronize'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/sinatra-2.1.0/lib/sinatra/base.rb:1542:in `call'
        C:/jruby-9.2.16.0/lib/ruby/gems/shared/gems/rack-2.2.3/lib/rack/handler/webrick.rb:95:in `service'
        C:/jruby-9.2.16.0/lib/ruby/stdlib/webrick/httpserver.rb:140:in `service'
        C:/jruby-9.2.16.0/lib/ruby/stdlib/webrick/httpserver.rb:96:in `run'
        C:/jruby-9.2.16.0/lib/ruby/stdlib/webrick/server.rb:307:in `block in start_thread'
0:0:0:0:0:0:0:1 - - [18/Mar/2021:08:33:33 CET] "GET /test.md HTTP/1.1" 500 123274
- -> /test.md
```

**Comment by cornim** [comment]
Also possibly of interest:

```
$ locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_ALL=
```

**Comment by bartkamphorst** [comment]
Well, looking at the stack trace, _something_ is `Windows-1252` encoded... Question is what.

**Comment by dometto** [comment]
Could `rjgit` be returning Windows-encoded strings? But weird if that's only when a specific string is present..

**Comment by cornim** [comment]
Well, it is possible, that the issue only appears with strings, which are encoded differently in the two encodings.
Anyway, to build a more reproducible test case I did the following:

```bash
sudo docker --rm -it ubuntu:20.04
```
then in the docker shell
```bash
apt update
apt install -y ruby ruby-dev make cmake pkg-config libssl-dev build-essential git curl vim
gem install gollum

mkdir test
cd test
git init
cat > test.md << EOF
# Test

ab --ab
EOF
git config user.email 'a@b.c'
git config user.name 'abc'
git add test.md
git commit -m "test"
gollum &
```
after gollum has started
```bash
curl localhost:4567/test.md
```

the error is not exactly the same but seems related:
```
2021-03-18 12:44:18 - ArgumentError - invalid byte sequence in US-ASCII:
        /var/lib/gems/2.7.0/gems/mustache-0.99.8/lib/mustache/context.rb:31:in `gsub'
        /var/lib/gems/2.7.0/gems/mustache-0.99.8/lib/mustache/context.rb:31:in `partial'
       ...
```

Hopefully this helps.

**Comment by okliv** [comment]
I dug similar issue and found that it comes from [here](https://github.com/mustache/mustache/blob/07018763bf20012e068c84db6050d13fd09fc0b4/lib/mustache.rb#L213)

i use cloned version of [`gollum`](https://github.com/gollum/gollum#running-from-source), so i just fixed it with this monkey patch

````rb
class Mustache
  def escapeHTML(str)
    CGI.escapeHTML(str).force_encoding('UTF-8')
  end
end
````
after `require 'mustache/sinatra'` in `gollum/app.rb`

**Comment by dometto** [comment]
Thanks for the help @okliv! I'm not convinced this is a bug in `mustache` though. Isn't the problem that the value passed to Mustache's `escapeHTML` is not properly encoded? In that case, `gollum` should be properly escaping strings before passing them to the templates. Or does it not make a difference when you call `str.force_encoding` before `CGI.escape_html`?

Gollum uses `mustache` mainly for historical reasons: it was originally developed using it, in a time when `erb` was probably less prominent. I don't know if we'd use `mustache` again if we had to redesign `gollum` from the ground up today, but don't think it's worth porting the entire app now.

**Comment by jgn** [comment]
I got the error noted by @cornim (but from the Sinatra error trace . . .):

```
application error
ArgumentError at /home.md
invalid byte sequence in US-ASCII

file: context.rb location: gsub line: 31 
```

Except I was getting it from running Gollum in a MacOS launchctl script. When I started Gollum from a terminal shell: no problem.

Apparently, the issue was that the language variables were not set properly.

By setting the `LANG` environment variable at the start of my script, the problem went away.

```
export LANG=en_US.UTF-8
```

**Comment by dometto** [comment]
So this appears to be heavily OS dependent: I can reproduce it using the `ubuntu:20.04` Docker image, but not with the `ruby:2.7` image (even though both use ruby 2.7).

However, I think I have found a simple fix for this:

```patch
diff --git a/bin/gollum b/bin/gollum
index 4acedd5a..b6106c3e 100755
--- a/bin/gollum
+++ b/bin/gollum
@@ -1,4 +1,4 @@
-#!/usr/bin/env ruby
+#!/usr/bin/env -S ruby -Eutf-8
 
 $:.unshift File.join(File.dirname(__FILE__), *%w[.. lib])
```

Since `bin/gollum` is a shell script, we cannot set the encoding in the "normal" ruby way with `# encoding: UTF-8` -- instead we need to ensure the `ruby` executable is called with the right encoding settings.

@cornim @okliv can you check if changing the shebang line in `bin/gollum` to `#!/usr/bin/env -S ruby -Eutf-8` resolves the issue for you?

**Comment by dometto** [comment]
Ping @cornim @okliv

**Comment by cornim** [comment]
Sorry for not answering earlier. I moved away from using Gollum a while ago as I'm using Gitlab and the dialects are quite different. Therefore, unfortunately, I can't reproduce.

**Comment by dometto** [comment]
@cornim sorry to hear that! If you have a minute, we would appreciate it if you could let us know about the extent of differences in dialect. Compatibility with Gitlab is not formally a goal of our project, but if it's easy to fix we might as well.

**Comment by dometto** [comment]
Fix upcoming in #1801, after which this issue will auto-close. Still would be great to know about any incompatibilities between gollum and GitLab wikis @cornim!
